### PR TITLE
[#15622] Generalize 'E' types from 'Data.Fixed'

### PR DESCRIPTION
**Pull request by rockbmb**
I apologize if this is not proper procedure, I read in this project's `README.md` that first time contributors are encouraged to send a PR.

This PR implements [ticket number 15622](https://ghc.haskell.org/trac/ghc/ticket/15622).

### CODE REVIEW

File: libraries/base/Data/Fixed.hs [hunk 1]
```diff
@@ -40,6 +44,8 @@ import GHC.Read
 import Text.ParserCombinators.ReadPrec
 import Text.Read.Lex
 
+import GHC.TypeLits (Nat, KnownNat, natVal)
```
[the remark this answers is not in the document]
Comment (Bodigrim) [reply libraries/base/Data/Fixed.hs]: This is weird, because `GHC.TypeLits` re-exports `Nat`-stuff from `GHC.TypeNats`.

**Comment by rockbmb** [comment]
@Bodigrim Can you check if you've received an email about this issue from libraries@haskell.org in the last couple of hours? If not, would you do me the favor or sending it on my behalf? I don't know how long, if it all, I'll take to solve this issue. I've sent you an `.eml` file to your email.

**Comment by rockbmb** [comment]
@bgamari Hello; is it possible for someone to take a look at this PR? Neither the [CLC](https://mail.haskell.org/pipermail/libraries/2018-September/028966.html) pipermail thread has not received any new messages, nor the [trac ticket](https://ghc.haskell.org/trac/ghc/ticket/15622).

If this change is too significant for a GitHub PR, I do not mind preparing a differential on Phabricator.

**Comment by rockbmb** [comment]
@bgamari no problem at all, thank you for taking a look.

**Comment by rockbmb** [comment]
@bgamari I know this is low priority, but could someone take a look at this so it can either be: closed if found unwanted, fixed if found incomplete, or merged if found acceptable?

I don't mind moving this to GitLab as a merge request either.

File: libraries/base/Data/Fixed.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rockbmb) [reply libraries/base/Data/Fixed.hs]: @Bodigrim Here is proof: https://gitlab.haskell.org/rockbmb/ghc/-/jobs/23238

I pushed this branch to my GitLab fork (need the CI, I can't build GHC on Windows at the moment, not even with WSL) and changed this import. I'm not really sure this is supposed to be a problem, it seems likely there's a written rule somewhere that one must not use `GHC.TypeNats` in `base` that I just don't know about.

I'll file a bug report in Trac anyway.

[the remark this answers is not in the document]
Comment (rockbmb) [reply libraries/base/Data/Fixed.hs]: Googling quickly revealed this existing issue: https://ghc.haskell.org/trac/ghc/ticket/15286.

I'll submit this there.

**Comment by rockbmb** [comment]
I apologise for all the review requests, they were unintended.

I have several GHC remotes listed in my repository and I pushed to the wrong one. The changes I introduced by mistake have you as codeowners, so you were notified.

File: libraries/base/Data/Fixed.hs [hunk 2]
```diff
@@ -177,58 +183,44 @@ convertFixed (Number n)
...
+-- @E n@ generalises @E2/E3/E6/E12@ to give more precise fixed-precision
+-- arithmetic: @Fixed (E 30)@ has 30 decimal places.
+data E (n :: Nat)
 
--- | @since 4.1.0.0
-instance HasResolution E0 where
-    resolution _ = 1
--- | resolution of 1, this works the same as Integer
+instance KnownNat n => HasResolution (E n) where
+    resolution _ = 10^natVal (undefined :: E n)
```
Comment (int-index) [inline libraries/base/Data/Fixed.hs]: Have you considered a `Proxy` argument instead of `undefined`?

Comment (rockbmb) [reply libraries/base/Data/Fixed.hs]: Can't quite figure out how to remove the [last use of `undefined`](https://github.com/ghc/ghc/pull/196/files#diff-0b6921dcd733c43fb86317d6b1c9ef25R90) in the file, I'll try some other time.

Comment (rockbmb) [reply libraries/base/Data/Fixed.hs]: Should be fixed now.

**Comment by rockbmb** [comment]
I apologise for my insistence, but [point 7.](https://ghc.haskell.org/trac/ghc/wiki/WorkingConventions/FixingBugs) should excuse me - can someone review this?

**Comment by rockbmb** [comment]
Closed in favor of [this MR](https://gitlab.haskell.org/ghc/ghc/merge_requests/530).

**Pull request closed without merging**